MDEV-12182 Add client tcp port to audit plugin logs

HashimSyedUBC opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-12182*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description
In environments with load balancers or proxies, the audit plugin logs
only the IP address, making it difficult to differentiate individual client
connections from the same IP.

Add a new 'port' field to the appropriate event objects to capture the
client's TCP port number. Populate the port field with thd->port  in the
appropriate functions. The audit plugin receives and logs this port field
along  with  other connection information, enabling better identification
of individual client connections.

## Release Notes
N/A

## How can this PR be tested?
- All existing tests pass when running ./mtr. 
- Execute `server_audit.test` file to verify to port logging feature.


## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch.*\

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3324) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3324) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3324) it.</sub>

LinuxJedi commented on plugin/server_audit/server_audit.c:
```diff
@@ -1401,6 +1403,14 @@ static size_t log_header(char *message, size_t message_len,
+    strncpy(port_str, "unavailable", sizeof(port_str) - 1);
+    port_str[sizeof(port_str) - 1] = '\0'; 
+  } else {
+    snprintf(port_str, sizeof(port_str) - 1, "%u", port); // Subtract 1 to account for null terminator
```
Probably best to put this comment above the `snprintf` to keep it under 80 lines. Also, please use `my_snprintf`.

HashimSyedUBC replied:
Removed the null termination and using my_snprintf now. Thank you.

LinuxJedi commented on plugin/server_audit/server_audit.c:
```diff
@@ -1401,6 +1403,14 @@ static size_t log_header(char *message, size_t message_len,
+    port_str[sizeof(port_str) - 1] = '\0'; 
+  } else {
+    snprintf(port_str, sizeof(port_str) - 1, "%u", port); // Subtract 1 to account for null terminator
+    port_str[sizeof(port_str) - 1] = '\0'; // Ensure null termination
```
Since the maximum possible length is 10 and you have allocated 16, this shouldn't be needed. Also `my_snprintf` would do this for you anyway.

HashimSyedUBC replied:
Removed the null termination and using my_snprintf now. Thank you.

LinuxJedi reviewed (commented):
I think it is a good feature to add, many thanks for the contribution.

I've added a couple of comments for some minor things. Also, the following things will be needed:

* MYSQL_AUDIT_INTERFACE_VERSION version will need to be bumped
* The version in `server_audit.c` will need to be bumped
* A MDEV will need to be attached to this, so it can be tracked through review / testing / documentation processes.

I can help with the last one of these if you wish.

HashimSyedUBC commented:
> I think it is a good feature to add, many thanks for the contribution.
> 
> I've added a couple of comments for some minor things. Also, the following things will be needed:
> 
> * MYSQL_AUDIT_INTERFACE_VERSION version will need to be bumped
> * The version in `server_audit.c` will need to be bumped
> * A MDEV will need to be attached to this, so it can be tracked through review / testing / documentation processes.
> 
> I can help with the last one of these if you wish.

Updated the minor version in server_audit.c and updated MYSQL_AUDIT_INTERFACE_VERSION. Also embedded the MDEV tickets in the first comment.

LinuxJedi reviewed (changes requested):
Many thanks for your fixes, unfortunately it is causing a compiler error now.

LinuxJedi commented on plugin/server_audit/server_audit.c:
```diff
@@ -1401,6 +1403,13 @@ static size_t log_header(char *message, size_t message_len,
     host= userip;
   }
+  char port_str[16];
+  if (port == NULL) {
```
Oh, also the second error is because we don't always compile with C99 standard, so the declaration needs to be at the top of the section.

HashimSyedUBC replied:
Yes I will keep this mind. I have relocated char port_str[16]; to the top of the function. Also changed the comparison from NULL to 0.

HashimSyedUBC commented:
Currently working on work-around for windows specific failing audit plugin test

HashimSyedUBC commented:
All failures have now been resolved. Please note that the failing tests were pre-existing and are not related to this recent change.

HashimSyedUBC commented:
#3425 All the failing tests are confirmed to be on main branch and not due to any changes made in this PR

vuvova commented:
The thing is, [MDEV-12182](https://jira.mariadb.org/browse/MDEV-12182) is marked as *blocked by* [MDEV-5313](https://jira.mariadb.org/browse/MDEV-5313) "Improving audit api". That is. instead if adding more fields ad hoc to audit plugin structures we wanted to consider a systematic approach to audit API extensions, and this was [MDEV-5313](https://jira.mariadb.org/browse/MDEV-5313). Which didn't go anywhere.

So, now we need to decide whether to add a new field ad hoc to audit plugin structures, just this one time as an exception or design the extensible audit API firts, as Jira says.

What is the compelling reason to make an exception here? Does anyone need a port in the audit log particularly badly?

vuvova commented:
Tomorrow's the last day, if there won't be a patch we can accept and you won't say "just fix it yourself", it will miss 12.0

vuvova commented on mysql-test/suite/plugins/t/server_audit.test:
```diff
@@ -237,8 +237,8 @@ show status like 'server_audit_current_log';
-# replace the timestamp and the hostname with constant values
---replace_regex /[0-9]* [0-9][0-9]:[0-9][0-9]:[0-9][0-9]\,[^,]*\,/TIME,HOSTNAME,/ /\,[1-9][0-9]*\,/,1,/ /\,[1-9][0-9]*/,ID/
+# replace the timestamp, hostname and port number with constant values
+--replace_regex /[0-9]* [0-9][0-9]:[0-9][0-9]:[0-9][0-9],[^,]*,/TIME,HOSTNAME,/ /,[1-9][0-9],/,1,/ /,[1-9][0-9]*/,ID/ /,localhost:([0-9]+|unavailable)/,localhost:port/
```
ah, sorry, that's not `$MASTER_MYPORT`

vuvova commented:
pushed into 12.0, thanks
